src/neo/SmartContract/InteropService.Contract.cs

```diff
@@ -25,6 +28,11 @@ private static long GetDeploymentPrice(EvaluationStack stack)
                 return Storage.GasPerByte * size;
             }
 
+            internal static Guid GetDeterministicGuid(Transaction tx, uint nonce)
+            {
+                return new Guid(Concat(tx.Hash.ToArray(), BitConverter.GetBytes(nonce)).Sha256().AsSpan(0, 16));
```

Comment:
Good, @erikzhang.

Reply:
Maybe it is not safe here.We can attack the storage area of ​​a contract by means of hash collision.

Reply:
Perhaps yes, @doubiliu, you are right.

```cs
using System;
using System.Runtime.InteropServices;

// Guid for the interface IMyInterface.
[Guid("F9168C5E-CEB2-4faa-B6BF-329BF39FA1E4")]
interface IMyInterface
{
    void MyMethod();
}

// Guid for the coclass MyTestClass.
[Guid("936DA01F-9ABD-4d9d-80C7-02AF85C822A8")]
public class MyTestClass : IMyInterface
{
    public void MyMethod() {}

    public static void Main( string []args )
    {
        GuidAttribute IMyInterfaceAttribute = (GuidAttribute) Attribute.GetCustomAttribute(typeof(IMyInterface), typeof(GuidAttribute));
        
        System.Console.WriteLine("IMyInterface Attribute: " + IMyInterfaceAttribute.Value );    

        // Use the string to create a guid.
        Guid myGuid1 = new Guid(IMyInterfaceAttribute.Value );
        // Use a byte array to create a guid.
        Guid myGuid2 = new Guid(myGuid1.ToByteArray());

        if (myGuid1.Equals(myGuid2))
            System.Console.WriteLine("myGuid1 equals myGuid2");
        else
            System.Console.WriteLine("myGuid1 does not equal myGuid2" );

        // Equality operator can also be used to determine if two guids have same value.
        if ( myGuid1 == myGuid2 )
            System.Console.WriteLine( "myGuid1 == myGuid2" );
        else
            System.Console.WriteLine( "myGuid1 != myGuid2" );
    }
}
// The example displays the following output:
//       IMyInterface Attribute: F9168C5E-CEB2-4faa-B6BF-329BF39FA1E4
//       myGuid1 equals myGuid2
//       myGuid1 == myGuid2
```
https://docs.microsoft.com/en-us/dotnet/api/system.guid?view=netframework-4.8

Reply:
So if we want to use Guid,we need design a safe generating way.And it must satisfy consistency

Reply:
We can use the `BlockHeight `+ `TransactionIndex `+ `SyscallCounter `to build the key.I think it is safe. @vncoelho @erikzhang

Reply:
I originally wanted to use TransactionIndex, but it was a bit cumbersome. `ApplicationEngine` needs to know the index of the transactions.

Reply:
Maybe you can modify this PR to use TransactionIndex in GUID.

Reply:
@doubiliu, can you go for it?

Reply:
@vncoelho @erikzhang  I can do it.But I think `global counter` will be a better way.  I will submit changes as soon as possible.

Reply:
That is true, @doubiliu, a globalcounter makes sense.